### PR TITLE
Support additional generic groups

### PR DESCRIPTION
**Pull request by bjonnh-work**
#### What does this implement/fix? Explain your changes.

Adding some missing search groups.

We are still a bit unhappy about the H ones as it would be nice if they could use the implicit hydrogens. We couldn't find an efficient way to do that, but we have been asked multiple times.

### CODE REVIEW

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 1]
```diff
@@ -238,42 +501,167 @@ RDKIT_GENERICGROUPS_EXPORT bool AlkoxyacyclicAtomMatcher(
...
+        {"HAH", Matchers::HeteroarylHAtomMatcher},
         {"NoCarbonRing", Matchers::NoCarbonRingAtomMatcher},
         {"CXX", Matchers::NoCarbonRingAtomMatcher},
+        {"NoCarbonRingH", Matchers::NoCarbonRingHAtomMatcher},
+        {"CXH", Matchers::NoCarbonRingHAtomMatcher},
+        {"D", Matchers::DAtomMatcher},
+        {"T", Matchers::TAtomMatcher},
+        {"H+", Matchers::HplusAtomMatcher},
+        {"Pol", Matchers::PolAtomMatcher},
+        {"R", Matchers::RAtomMatcher},
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:662]: I think these are all "normal" atoms and should be handled as such. They aren't really queries.
What's the motivation for adding them here?

Comment (bjonnh-work) [reply Code/GraphMol/GenericGroups/GenericGroups.h:662]: I don't think we use these inside sgroups, so we can probably remove them from that PR.

Comment (tadhurst-cdd) [reply Code/GraphMol/GenericGroups/GenericGroups.h]: These are on the list of expanded groups inside Ketcher editor

File: Docs/Book/RDKit_Book.rst [hunk 2]
```diff
@@ -1547,21 +1547,46 @@ Here's an example of using the features:
...
+  HeterocyclicH (CHH)       cyclic side chains with at least one heteroatom or an H atom
   Heteroaryl (HAR)          aryl side chains with at least one heteroatom
+  HeteroarylH (HAH)         aryl side chains with at least one heteroatom or an H atom
   NoCarbonRing (CXX)        ring containing no carbon atoms
+  NoCarbonRingH (CXH)       ring containing no carbon atoms or an H atom
+  Group (G)                 any group (not H atom)
+  GroupH (GH)               any group (including H atom)
+  Group* (G*)               any group with a ring closure
+  GroupH* (GH*)             any group with a ring closure or an H atom
+  Deuterium (D)             a deuterium atom
```
Comment (greglandrum) [inline Docs/Book/RDKit_Book.rst:1584]: I think the ones from here down are no longer relevant

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 3]
```diff
@@ -33,6 +33,60 @@ namespace GenericGroups {
 // I've tried to document deviations or surprises
 
 namespace Matchers {
+
+//! Matches any group as a side chains
+/*!
+
+  Note: this is Reaxys query type G and matches any sidechains
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:40]:
```suggestion
  Note: this is Reaxys query type G and matches any sidechain
```

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 4]
```diff
@@ -33,6 +33,60 @@ namespace GenericGroups {
 // I've tried to document deviations or surprises
 
 namespace Matchers {
+
+//! Matches any group as a side chains
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:37]:
```suggestion
//! Matches any group as a side chain
```

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 5]
```diff
@@ -33,6 +33,60 @@ namespace GenericGroups {
...
+  Note: this is Reaxys query type G and matches any sidechains
+
+  Conditions:
+    - at least one non-hydrogen atom is in the sidechain
+
+*/
+RDKIT_GENERICGROUPS_EXPORT bool GroupAtomMatcher(
+    const ROMol &mol, const Atom &atom, boost::dynamic_bitset<> ignore);
+
+//! Matches any group as a side chains including just an H atom
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:49]:
```suggestion
//! Matches any group as a side chain including just an H atom
```

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 6]
```diff
@@ -33,6 +33,60 @@ namespace GenericGroups {
...
+    - none
+
+*/
+RDKIT_GENERICGROUPS_EXPORT bool GroupHAtomMatcher(
+    const ROMol &mol, const Atom &atom, boost::dynamic_bitset<> ignore);
+
+//! Matches any group as a side chains
+/*!
+
+  Note: this is Reaxys query type G and matches any sidechain that has a ring
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:64]:
```suggestion
  Note: this is Reaxys query type G* and matches any sidechain that has a ring
```

File: Code/GraphMol/GenericGroups/GenericGroups.h [hunk 7]
```diff
@@ -33,6 +33,60 @@ namespace GenericGroups {
...
+
+  Conditions:
+    - at least one non-hydrogen atom is in the sidechain
+    - at least one ring closure
+
+*/
+RDKIT_GENERICGROUPS_EXPORT bool GroupStarAtomMatcher(
+    const ROMol &mol, const Atom &atom, boost::dynamic_bitset<> ignore);
+
+//! Matches any group as a side chains that has a ring closure or just an H atom
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/GenericGroups.h:75]:
```suggestion
//! Matches any group as a side chain that has a ring closure or just an H atom
```

File: Code/GraphMol/GenericGroups/generic_tests.cpp [hunk 8]
```diff
@@ -609,3 +218,520 @@ TEST_CASE("generics are compatible with extraFinalChecks",
...
+    REQUIRE(mol);
+    CHECK_THAT(*query, IsSubstructOf(*mol, 0, ps));
+  }
+}
+
+void runCxsmilesTest(const std::string groupToTest,
+                     std::vector<std::pair<std::string, unsigned>> &tests) {
+  std::string queryString = getCXSmilesQuery(groupToTest);
+  auto query = std::unique_ptr<RWMol>(SmartsToMol(queryString));
+  runTest(&*query, tests);
```
Comment (greglandrum) [inline Code/GraphMol/GenericGroups/generic_tests.cpp:304]:
```suggestion
  runTest(query.get(), tests);
```

**Review by greglandrum — CHANGES_REQUESTED** [review]
Mainly small stuff here

File: Code/GraphMol/GenericGroups/generic_tests.cpp [hunk 9]
```diff
@@ -609,3 +218,520 @@ TEST_CASE("generics are compatible with extraFinalChecks",
...
+void runMolTest(const std::string groupToTest,
+                std::vector<std::pair<std::string, unsigned>> &tests,
+                bool supGroupFlag) {
+  std::string queryString = getMolQuery(groupToTest, supGroupFlag);
+  auto query =
+      std::unique_ptr<RWMol>(MolBlockToMol(queryString, false, false, false));
+  runTest(&*query, tests);
+}
+
+void runTest(const std::string groupToTest,
```
[the remark this answers is not in the document]
Comment (tadhurst-cdd) [reply Code/GraphMol/GenericGroups/generic_tests.cpp:324]: Added.

**Comment by greglandrum** [comment]
@ricrogz @ptosco @bp-kelley
I think this PR mostly looks good, but since it results in a change in behavior I wanted to check with you on one point to see if you have strong objections.

Current behavior: if you want to include a generic group like ARY or AOX in a mol block, you have to include it as an S-Group.
Behavior after this PR: you can include ARY or AOX as atom labels, so this becomes a legal mol block:
```

  Mrv2211 08032316552D          

  0  0  0     0  0            999 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 2 1 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -6.5833 1.9167 0 0
M  V30 2 ARY -5.2497 2.6867 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 END BOND
M  V30 END CTAB
M  END
```
I think this is fine, and it certainly makes interoperability with chemical sketchers like marvin and ketcher easier.
Any objections to this change?

**Comment by ptosco** [comment]
> Any objections to this change?

I have no objections.

**Comment by ricrogz** [comment]
>  Any objections to this change?

No objections either, although I haven't used the generic groups. I think @rachelnwalker has, so maybe she think of something :)

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**